certbot/hooks.py

```diff
@@ -36,35 +44,51 @@ def validate_hook(shell_cmd, hook_name):
 
 def pre_hook(config):
     "Run pre-hook if it's defined and hasn't been run."
-    if config.pre_hook and not pre_hook.already:
-        logger.info("Running pre-hook command: %s", config.pre_hook)
-        _run_hook(config.pre_hook)
-    pre_hook.already = True
+    cmd = config.pre_hook
+    if cmd and cmd not in pre_hook.already:
+        logger.info("Running pre-hook command: %s", cmd)
+        _run_hook(cmd)
+        pre_hook.already[cmd] = True
+    elif cmd:
+        logger.info("Pre-hook command already run, skipping: %s", cmd)
 
-pre_hook.already = False
+pre_hook.already = {}
 
-def post_hook(config, final=False):
+
+def post_hook(config):
     """Run post hook if defined.
 
     If the verb is renew, we might have more certs to renew, so we wait until
-    we're called with final=True before actually doing anything.
+    run_saved_post_hooks() is called.
     """
-    if config.post_hook:
-        if not pre_hook.already:
-            logger.info("No renewals attempted, so not running post-hook")
-            if config.verb != "renew":
-                logger.warning("Sanity failure in renewal hooks")
-            return
-        if final or config.verb != "renew":
-            logger.info("Running post-hook command: %s", config.post_hook)
-            _run_hook(config.post_hook)
+
+    cmd = config.post_hook
+    # In the "renew" case, we save these up to run at the end
+    if config.verb == "renew":
+        if cmd and cmd not in post_hook.eventually:
+            post_hook.eventually.append(cmd)
+    # certonly / run
+    elif cmd:
+        logger.info("Running post-hook command: %s", cmd)
+        _run_hook(cmd)
+
+post_hook.eventually = []
+
+def run_saved_post_hooks():
+    """Run any post hooks that were saved up in the course of the 'renew' verb"""
+    for cmd in post_hook.eventually:
+        logger.info("Running post-hook command: %s", cmd)
+        _run_hook(cmd)
+    if len(post_hook.eventually) == 0:
```

Comment:
Posting this again here for tracking purposes because the relevant lines changed on the old comment.

nit: From your favorite document https://www.python.org/dev/peps/pep-0008/#id49
```
For sequences, (strings, lists, tuples), use the fact that empty sequences are false.

Yes: if not seq:
     if seq:

No: if len(seq):
    if not len(seq):
```